pulsar-client/src/main/java/org/apache/pulsar/client/impl/PulsarClientImpl.java

```diff
@@ -982,6 +1000,10 @@ public CompletableFuture<ClientCnx> getConnection(final String topic, final Stri
 
     public LookupService getLookup(String serviceUrl) {
         return urlLookupMap.computeIfAbsent(serviceUrl, url -> {
+            if (isClosed()) {
+                throw new IllegalStateException("Pulsar client has been closed, can not build LookupService when"
+                        + " calling get lookup with an url");
+            }
```

Comment:
Move it before the `computeIfAbsent`.

Actually it should be an internal API. If users don't call it by cast a `PulsarClient` object to `PulsarClientImpl`, this case should not happen.

A better solution is to make it package visible, as well as the `getConnection` method.

Reply:
> I don't think it's actual a memory leak. This PR only clears the map after closeAsync is called. However, I could hardly think of a case that a PulsarClient is still used after closeAsync. When the PulsarClientImpl object is GC'ed, this map will be GC'ed as well.

- The map obj is not an important point.
- `LookupService` is a child class of `Closable`, which means it always holds resources (such as connections or other resources). Before lookup services have been GC, should call `close`. 

> Move it before the `computeIfAbsent`.

In the method `computeIfAbsent` is better, because there is a case like below:
| `thread-1`| `thread-2` |
| --- | --- |
| for-each: close |
| | computeIfAbsent |

The new lookup service created through `thread-2` will not be released.

Reply:
> Threads get leaked since the underlying HttpClient of the HttpLookupService doesn't get closed.

@lhotari I get the point now. Yeah without this patch the connections won't be closed gracefully.

> In the method computeIfAbsent is better, because there is a case like below:

@poorbarcode It makes sense. Here you leverages the synchronized mechanism of `computeIfAbsent` to prevent the race.